In [1]:
# Importa las librerías necesarias y carga tus datos
import pandas as pd
import numpy as np


In [2]:


# Abre el archivo CSV
data_items = pd.read_csv(r'C:\Users\juana\Desktop\Cursos Python\HENRY\Proyecto 1\DATASET\datos_items.csv')

# Muestra el contenido del DataFrame
print(data_items)


                       user_id  items_count           steam_id  \
0            76561197970982479          277  76561197970982479   
1                      js41637          888  76561198035864385   
2                    evcentric          137  76561198007712555   
3                   Riot-Punch          328  76561197963445855   
4                        doctr          541  76561198002099482   
...                        ...          ...                ...   
87621        76561198323066619           22  76561198323066619   
87622        76561198326700687          177  76561198326700687   
87623  XxLaughingJackClown77xX            0  76561198328759259   
87624        76561198329548331            7  76561198329548331   
87625        edward_tremethick            0  76561198331598578   

                                                   items  playtime_forever  \
0      [{'item_id': '10', 'item_name': 'Counter-Strik...                 6   
1      [{'item_id': '10', 'item_name': 'Counter-Str

In [3]:
# Abre el archivo CSV
data_games = pd.read_csv(r'C:\Users\juana\Desktop\Cursos Python\HENRY\Proyecto 1\DATASET\limpio_games.csv')

# Muestra el contenido del DataFrame
print(data_games)


          publisher      genres             app_name                title  \
0         Kotoshiro      Action  Lost Summoner Kitty  Lost Summoner Kitty   
1         Kotoshiro      Casual  Lost Summoner Kitty  Lost Summoner Kitty   
2         Kotoshiro       Indie  Lost Summoner Kitty  Lost Summoner Kitty   
3         Kotoshiro  Simulation  Lost Summoner Kitty  Lost Summoner Kitty   
4         Kotoshiro    Strategy  Lost Summoner Kitty  Lost Summoner Kitty   
...             ...         ...                  ...                  ...   
57577  Laush Studio       Indie        Russian Roads        Russian Roads   
57578  Laush Studio      Racing        Russian Roads        Russian Roads   
57579  Laush Studio  Simulation        Russian Roads        Russian Roads   
57580      SIXNAILS      Casual  EXIT 2 - Directions  EXIT 2 - Directions   
57581      SIXNAILS       Indie  EXIT 2 - Directions  EXIT 2 - Directions   

                                                     url release_date  \
0 

In [4]:
# Abre el archivo CSV
data_reviews = pd.read_csv(r'C:\Users\juana\Desktop\Cursos Python\HENRY\Proyecto 1\DATASET\limpio_reviews.csv')

# Muestra el contenido del DataFrame
print(data_reviews)

                 user_id  item_id  \
0      76561197970982479     1250   
1      76561197970982479    22200   
2      76561197970982479    43110   
3                js41637   251610   
4                js41637   227300   
...                  ...      ...   
59300  76561198312638244       70   
59301  76561198312638244   362890   
59302        LydiaMorley   273110   
59303        LydiaMorley      730   
59304        LydiaMorley      440   

                                               helpful  recommend  \
0                                       No ratings yet       True   
1                                       No ratings yet       True   
2                                       No ratings yet       True   
3      15 of 20 people (75%) found this review helpful       True   
4         0 of 1 people (0%) found this review helpful       True   
...                                                ...        ...   
59300                                   No ratings yet       True   
593

In [5]:

# Renombrar la columna 'id' a 'item_id' en data_games
data_games.rename(columns={'id': 'item_id'}, inplace=True)

# Ahora la columna 'id' se llama 'item_id'

# Puedes verificar el cambio imprimiendo los nombres de las columnas
print(data_games.columns)


# Supongamos que tienes tus DataFrames data_games y data_reviews con 'id' renombrada a 'item_id'.

# Realiza la fusión (merge) en función de la columna 'item_id'
resulta = data_reviews.merge(data_games, on='item_id', how='inner')
resulta


Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'item_id', 'developer'],
      dtype='object')


user_id  item_id  \
0       76561197970982479     1250   
1            death-hunter     1250   
2                DJKamBer     1250   
3               diego9031     1250   
4       76561198081962345     1250   
...                   ...      ...   
123475       MeloncraftLP   262850   
123476         vinquility   431510   
123477         vinquility   431510   
123478         vinquility   431510   
123479         vinquility   431510   

                                              helpful  recommend  \
0                                      No ratings yet       True   
1                                      No ratings yet       True   
2                                      No ratings yet       True   
3                                      No ratings yet       True   
4                                      No ratings yet       True   
...                                               ...        ...   
123475   0 of 1 people (0%) found this review helpful       True   
123476  8 of 9 people (89%) found this review helpful       True   
123477  8 of 9 people (89%) found this review helpful       True   
123478  8 of 9 people (89%) found this review helpful       True   
123479  8 of 9 people (89%) found this review helpful       True   

                                                   review  year_posted  \
0       Simple yet with great replayability. In my opi...         2011   
1       Amazing, Non-stop action of blowing stuff to b...         2015   
2       Compared to Left 4 Dead 2, this game REALLY gi...         2013   
3                                              Jogo ♥♥♥♥.         2015   
4       cara nas imagens esse jogo da pouco de medo ma...         2014   
...                                                   ...          ...   
123475  THIS GAME!!!!!!!!!!!!!!!!!! WOOOOOOOOOOOOOOOOO...         2013   
123476  Highly recommended! The storyline for shou is ...         2013   
123477  Highly recommended! The storyline for shou is ...         2013   
123478  Highly recommended! The storyline for shou is ...         2013   
123479  Highly recommended! The storyline for shou is ...         2013   

        sentiment_analysis             publisher      genres  \
0                        2  Tripwire Interactive      Action   
1                        2  Tripwire Interactive      Action   
2                        0  Tripwire Interactive      Action   
3                        2  Tripwire Interactive      Action   
4                        2  Tripwire Interactive      Action   
...                    ...                   ...         ...   
123475                   2             SkyGoblin       Indie   
123476                   2    Aeon Dream Studios   Adventure   
123477                   2    Aeon Dream Studios      Casual   
123478                   2    Aeon Dream Studios       Indie   
123479                   2    Aeon Dream Studios  Simulation   

                                      app_name  \
0                                Killing Floor   
1                                Killing Floor   
2                                Killing Floor   
3                                Killing Floor   
4                                Killing Floor   
...                                        ...   
123475           The Journey Down: Chapter Two   
123476  Mystic Destinies: Serendipity of Aeons   
123477  Mystic Destinies: Serendipity of Aeons   
123478  Mystic Destinies: Serendipity of Aeons   
123479  Mystic Destinies: Serendipity of Aeons   

                                         title  \
0                                Killing Floor   
1                                Killing Floor   
2                                Killing Floor   
3                                Killing Floor   
4                                Killing Floor   
...                                        ...   
123475           The Journey Down: Chapter Two   
123476  Mystic Destinies: Serendipity of Aeons   
123477  Mystic Destinies: Serendipi

In [6]:
data_games.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'item_id', 'developer'],
      dtype='object')

In [7]:
data_items.columns

Index(['user_id', 'items_count', 'steam_id', 'items', 'playtime_forever',
       'playtime_2weeks', 'item_id'],
      dtype='object')

In [8]:
data_reviews.columns

Index(['user_id', 'item_id', 'helpful', 'recommend', 'review', 'year_posted',
       'sentiment_analysis'],
      dtype='object')

In [9]:
resulta.columns

Index(['user_id', 'item_id', 'helpful', 'recommend', 'review', 'year_posted',
       'sentiment_analysis', 'publisher', 'genres', 'app_name', 'title', 'url',
       'release_date', 'tags', 'reviews_url', 'specs', 'price', 'developer'],
      dtype='object')

ENDPOINT 5

In [23]:
# Crea un nuevo DataFrame con las columnas que necesitas
nuevo_resulta = resulta[['sentiment_analysis', 'developer']]

# Imprime las primeras filas del nuevo DataFrame para verificar
nuevo_resulta.columns


Index(['sentiment_analysis', 'developer'], dtype='object')

In [24]:
resultado = nuevo_resulta 

In [26]:

# Guardar el DataFrame 'result' en un archivo CSV
resultado.to_csv('resultado_endcinco.csv', index=False)


In [34]:

# Cargar el archivo CSV en un nuevo DataFrame
resultado = pd.read_csv('resultado_endcinco.csv')

# Define la función que toma el nombre del desarrollador como entrada
def developer_reviews_analysis(desarrolladora):
    # Filtra las reseñas para el desarrollador especificado
    developer_data = resultado[resultado['developer'] == desarrolladora]

    # Contar la cantidad de reseñas positivas, negativas y neutras para ese desarrollador
    total_reviews = len(developer_data)
    positive_reviews = (developer_data['sentiment_analysis'] == 2).sum()
    negative_reviews = (developer_data['sentiment_analysis'] == 0).sum()
    neutral_reviews = (developer_data['sentiment_analysis'] == 1).sum()

    # Convertir los resultados a enteros
    positive_reviews = int(positive_reviews)
    negative_reviews = int(negative_reviews)
    neutral_reviews = int(neutral_reviews)

    # Devuelve la información del desarrollador como un diccionario
    analysis_result = {
        "desarrolladora": desarrolladora,
        "total_resenias": total_reviews,
        "positive_reviews": positive_reviews,
        "negative_reviews": negative_reviews,
        "neutral_reviews": neutral_reviews
    }

    # Retorna el resultado
    return analysis_result

# Llama a la función con el nombre del desarrollador que desees analizar
resultado_analisis = developer_reviews_analysis('Valve')

# Imprime el resultado
print(resultado_analisis)


{'desarrolladora': 'Valve', 'total_resenias': 13920, 'positive_reviews': 12441, 'negative_reviews': 1465, 'neutral_reviews': 14}


In [32]:


# Define la función que toma el nombre del desarrollador como entrada
def developer_reviews_analysis(desarrolladora):
    # Filtra las reseñas para el desarrollador especificado
    developer_data = resultado[resultado['developer'] == desarrolladora]

    # Contar la cantidad de reseñas positivas, negativas y neutras para ese desarrollador
    total_reviews = len(developer_data)
    positive_reviews = (developer_data['sentiment_analysis'] == 2).sum()
    negative_reviews = (developer_data['sentiment_analysis'] == 0).sum()
    neutral_reviews = (developer_data['sentiment_analysis'] == 1).sum()



    # Imprime la información sobre el desarrollador
    print(f'Análisis de reseñas para el desarrollador {desarrolladora}:')
    print(f'Total de reseñas: {total_reviews}')
    print(f'Reseñas positivas: {positive_reviews}')
    print(f'Reseñas negativas: {negative_reviews}')
    print(f'Reseñas neutras: {neutral_reviews}')

# Llama a la función con el nombre del desarrollador que desees analizar
developer_reviews_analysis('Valve')


Análisis de reseñas para el desarrollador Valve:
Total de reseñas: 13920
Reseñas positivas: 12441
Reseñas negativas: 1465
Reseñas neutras: 14


In [15]:

import random

# Obtiene una lista única de nombres de desarrolladores
desarrolladores_unicos = resulta['developer'].unique()

# Selecciona aleatoriamente 10 nombres de desarrolladores
nombres_de_desarrolladores = random.sample(list(desarrolladores_unicos), 10)

# Imprime los 10 nombres de desarrolladores
print("10 nombres de desarrolladores aleatorios:")
for nombre in nombres_de_desarrolladores:
    print(nombre)


10 nombres de desarrolladores aleatorios:
ACQUIRE Corp.
773
KintoGames
Noah System,Mgame
Fistful of Frags Team
S2 Games, LLC
Galactic Cafe
Odin Game Studio
MOONSTONE
Sluggerfly


ENDPOINT 4

data_reviews: Contiene información sobre las revisiones, y necesitaremos las columnas user_id, item_id, recommend (para verificar si es una revisión recomendada) y sentiment_analysis (para determinar si los comentarios son positivos).

data_games: Contiene información sobre los juegos, y necesitaremos las columnas id (para vincular los juegos con las revisiones) y developer (para identificar a los desarrolladores de los juegos).

In [20]:

# Renombrar la columna 'id' a 'item_id' en data_games
data_games.rename(columns={'id': 'item_id'}, inplace=True)

# Ahora la columna 'id' se llama 'item_id'

# Puedes verificar el cambio imprimiendo los nombres de las columnas
print(data_games.columns)


# Supongamos que tienes tus DataFrames data_games y data_reviews con 'id' renombrada a 'item_id'.

# Realiza la fusión (merge) en función de la columna 'item_id'
result = data_reviews.merge(data_games, on='item_id', how='inner')



result.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'item_id', 'developer'],
      dtype='object')


Index(['user_id', 'item_id', 'helpful', 'recommend', 'review', 'year_posted',
       'sentiment_analysis', 'publisher', 'genres', 'app_name', 'title', 'url',
       'release_date', 'tags', 'reviews_url', 'specs', 'price', 'developer'],
      dtype='object')

In [23]:
result.columns

Index(['user_id', 'item_id', 'helpful', 'recommend', 'review', 'year_posted',
       'sentiment_analysis', 'publisher', 'genres', 'app_name', 'title', 'url',
       'release_date', 'tags', 'reviews_url', 'specs', 'price', 'developer'],
      dtype='object')

In [36]:
# Lista de columnas necesarias
columnas_necesarias = ['user_id', 'item_id', 'year_posted', 'recommend', 'sentiment_analysis', 'developer']

# Filtrar el DataFrame 'result' para mantener solo las columnas necesarias
result = result[columnas_necesarias]
result 

user_id  item_id  year_posted  recommend  \
0       76561197970982479     1250         2011       True   
1            death-hunter     1250         2015       True   
2                DJKamBer     1250         2013       True   
3               diego9031     1250         2015       True   
4       76561198081962345     1250         2014       True   
...                   ...      ...          ...        ...   
123475       MeloncraftLP   262850         2013       True   
123476         vinquility   431510         2013       True   
123477         vinquility   431510         2013       True   
123478         vinquility   431510         2013       True   
123479         vinquility   431510         2013       True   

        sentiment_analysis             developer  
0                        2  Tripwire Interactive  
1                        2  Tripwire Interactive  
2                        0  Tripwire Interactive  
3                        2  Tripwire Interactive  
4                        2  Tripwire Interactive  
...                    ...                   ...  
123475                   2             SkyGoblin  
123476                   2    Aeon Dream Studios  
123477                   2    Aeon Dream Studios  
123478                   2    Aeon Dream Studios  
123479                   2    Aeon Dream Studios  

[123480 rows x 6 columns]

In [38]:
# Guardar el DataFrame 'result' en un archivo CSV
result.to_csv('resultados_endcuatro.csv', index=False)


In [39]:

# Cargar el archivo CSV en un nuevo DataFrame
nuevo_resultado = pd.read_csv('resultados_endcuatro.csv')

# Imprimir el nuevo DataFrame para verificar
print(nuevo_resultado)


                  user_id  item_id  year_posted  recommend  \
0       76561197970982479     1250         2011       True   
1            death-hunter     1250         2015       True   
2                DJKamBer     1250         2013       True   
3               diego9031     1250         2015       True   
4       76561198081962345     1250         2014       True   
...                   ...      ...          ...        ...   
123475       MeloncraftLP   262850         2013       True   
123476         vinquility   431510         2013       True   
123477         vinquility   431510         2013       True   
123478         vinquility   431510         2013       True   
123479         vinquility   431510         2013       True   

        sentiment_analysis             developer  
0                        2  Tripwire Interactive  
1                        2  Tripwire Interactive  
2                        0  Tripwire Interactive  
3                        2  Tripwire Interactive  


In [41]:

# Cargar el archivo CSV en un nuevo DataFrame
result = pd.read_csv('resultados_endcuatro.csv')


def mejores_desarrolladores_anio(anio: int):
    # Paso 1: Filtrar las revisiones para el año proporcionado y donde recommend es True y sentiment_analysis es positivo
    revisiones_filtradas = result[(result['year_posted'] == anio) &
                                  (result['recommend'] == True) &
                                  (result['sentiment_analysis'] > 0)]

    # Paso 2: Agrupar por desarrollador y contar la cantidad de juegos recomendados
    desarrolladores_recomendados = revisiones_filtradas.groupby('developer')['recommend'].count().reset_index()

    # Paso 3: Ordenar en orden descendente y tomar el top 3
    top_desarrolladores = desarrolladores_recomendados.sort_values(by='recommend', ascending=False).head(3)

    # Verificar si top_desarrolladores tiene al menos 3 elementos
    if len(top_desarrolladores) >= 3:
        resultado = [{"Puesto " + str(i + 1): top_desarrolladores.iloc[i]['developer']} for i in range(3)]
    else:
        resultado = [{"Puesto " + str(i + 1): "No disponible"} for i in range(len(top_desarrolladores))]

    return resultado

resultado = mejores_desarrolladores_anio(2015)
print(resultado)


[{'Puesto 1': 'Valve'}, {'Puesto 2': 'Facepunch Studios'}, {'Puesto 3': 'Smartly Dressed Games'}]


In [33]:
# Muestra los datos de la columna 'year_posted' en el DataFrame 'result'
datos_año_posted = result['year_posted']

# Imprime los primeros 10 valores para visualizarlos
print(datos_año_posted.head(10))


0    2011
1    2015
2    2013
3    2015
4    2014
5    2010
6    2014
7    2015
8    2014
9    2014
Name: year_posted, dtype: int64


ENDPOINT 3

In [97]:

# Crear un nuevo DataFrame a partir de data_reviews con las columnas seleccionadas
data_reviews_nuevo = data_reviews[['user_id', 'item_id', 'year_posted', 'sentiment_analysis']]

# Crear un nuevo DataFrame a partir de data_games con las columnas seleccionadas
data_games_nuevo = data_games[['app_name', 'release_date', 'genres', 'id']]

# Guardar el DataFrame data_reviews_nuevo en un archivo CSV
data_reviews_nuevo.to_csv('data_reviews_endtres.csv', index=False)

# Guardar el DataFrame data_games_nuevo en un archivo CSV
data_games_nuevo.to_csv('data_games_endtres.csv', index=False)


In [96]:
# Cargar el archivo CSV de data_reviews_nuevo
data_reviews_nuevo = pd.read_csv('data_reviews_endtres.csv')


# Cargar el archivo CSV de data_games_nuevo
data_games_nuevo = pd.read_csv('data_games_endtres.csv')


def UserForGenre(data_games_nuevo, data_reviews_nuevo, genre):
    # Paso 1: Filtrar los juegos del género específico en data_games
    genre_filter = data_games_nuevo['genres'].str.contains(genre, case=False, na=False)
    genre_games = data_games_nuevo[genre_filter]

    if genre_games.empty:
        return f"No se encontraron juegos en el género '{genre}'"

    # Paso : Filtrar las reseñas en data_reviews que corresponden a estos juegos
    genre_reviews = data_reviews_nuevo[data_reviews_nuevo['item_id'].isin(genre_games['id'])]

    # Paso : Calcular la cantidad total de horas jugadas por año de lanzamiento en las reseñas
    playtime_by_year = genre_reviews.groupby('year_posted')['sentiment_analysis'].count().reset_index()
    playtime_by_year.columns = ['Año', 'Horas']

    # Paso : Encontrar el usuario con más horas jugadas y construir el resultado deseado
    max_user_id = genre_reviews.groupby('user_id')['sentiment_analysis'].count().idxmax()
    
    result = {
        "Usuario con más horas jugadas para género": max_user_id,
        "Horas jugadas": playtime_by_year.to_dict(orient='records')
    }

    return result

# Uso de la función
genero = "Action"  # Reemplaza "Action" con el género que estás buscando
resultado = UserForGenre(data_games_nuevo, data_reviews_nuevo, genero)
print(resultado)


{'Usuario con más horas jugadas para género': '76561198064484479', 'Horas jugadas': [{'Año': 2010, 'Horas': 36}, {'Año': 2011, 'Horas': 347}, {'Año': 2012, 'Horas': 848}, {'Año': 2013, 'Horas': 10461}, {'Año': 2014, 'Horas': 14078}, {'Año': 2015, 'Horas': 11295}]}


ENDPOINT 2

In [ ]:
# Crear un nuevo DataFrame con las columnas de interés
user_data_subset = pd.DataFrame({
    'user_id': data_reviews['user_id'],
    'recommend': data_reviews['recommend'],
    'item_id': data_items['item_id'],
    'price': data_games['price'],
    # Agrega más columnas según tus necesidades
})

# Guardar el nuevo DataFrame en un archivo CSV
#user_data_subset.to_csv('user_data_subset.csv', index=False)


In [ ]:
# Abre el archivo CSV
data_dos = pd.read_csv('user_data_subset.csv')

# Muestra el contenido del DataFrame
print(data_dos)

                 user_id recommend  item_id  price
0      76561197970982479      True       10   4.99
1      76561197970982479      True       10   4.99
2      76561197970982479      True     1200   4.99
3                js41637      True       10   4.99
4                js41637      True      300   4.99
...                  ...       ...      ...    ...
87621                NaN       NaN   413850    NaN
87622                NaN       NaN    11020    NaN
87623                NaN       NaN        0    NaN
87624                NaN       NaN   304930    NaN
87625                NaN       NaN        0    NaN

[87626 rows x 4 columns]


In [ ]:
from typing import Dict

def userdata(user_id: str, data_dos):
    # Filtrar las revisiones del usuario
    user_reviews = data_dos[data_dos['user_id'] == user_id]

    # Filtrar las compras del usuario
    user_purchases = data_dos[data_dos['user_id'] == user_id]

    # Calcular la cantidad total de dinero gastado por el usuario
    user_game_ids = user_purchases['item_id']
    total_money_spent = user_purchases['price'].sum()

    # Calcular el porcentaje de recomendación promedio
    total_recommendation_percentage = user_reviews['recommend'].mean()

    # Contar la cantidad de elementos comprados por el usuario
    num_items_purchased = len(user_purchases)

    user_data = {
        "Usuario": user_id,
        "Dinero gastado": f"{total_money_spent} USD",
        "% de recomendación": f"{total_recommendation_percentage:.2f}%",
        "Cantidad de items": num_items_purchased,
    }

    return user_data
result = userdata("76561197970982479", data_dos)
print(result)


{'Usuario': '76561197970982479', 'Dinero gastado': '14.97 USD', '% de recomendación': '1.00%', 'Cantidad de items': 3}


In [ ]:
from typing import Dict

def userdata(user_id: str, data_reviews, data_items, data_games):
    # Filtrar las revisiones del usuario
    user_reviews = data_reviews[data_reviews['user_id'] == user_id]

    # Filtrar las compras del usuario
    user_purchases = data_items[data_items['user_id'] == user_id]

    # Calcular la cantidad total de dinero gastado por el usuario
    user_game_ids = user_purchases['item_id']
    total_money_spent = data_games[data_games['id'].isin(user_game_ids)]['price'].sum()

    # Calcular el porcentaje de recomendación promedio
    total_recommendation_percentage = user_reviews['recommend'].mean()

    # Contar la cantidad de elementos comprados por el usuario
    num_items_purchased = len(user_purchases)

    user_data = {
        "Usuario": user_id,
        "Dinero gastado": f"{total_money_spent} USD",
        "% de recomendación": f"{total_recommendation_percentage:.2f}%",
        "Cantidad de items": num_items_purchased,
    }

    return user_data
result = userdata("76561197970982479", data_reviews, data_items, data_games)
print(result)



{'Usuario': '76561197970982479', 'Dinero gastado': '9.99 USD', '% de recomendación': '1.00%', 'Cantidad de items': 1}


ENDPOINT 1

In [ ]:
# Crear un nuevo DataFrame con las columnas seleccionadas
nuevo_data_games = data_games[['id','price', 'release_date', 'app_name', 'developer']].copy()

# Verificar el nuevo DataFrame
data_games = nuevo_data_games 

#data_games.to_csv('limpio_games_gituno.csv', index=False) 

In [ ]:


# Abre el archivo CSV
data_gamesgit = pd.read_csv(r'limpio_games_gituno.csv')

# Muestra el contenido del DataFrame
print(data_gamesgit)

           id  price release_date             app_name  \
0      761140   4.99   2018-01-04  Lost Summoner Kitty   
1      761140   4.99   2018-01-04  Lost Summoner Kitty   
2      761140   4.99   2018-01-04  Lost Summoner Kitty   
3      761140   4.99   2018-01-04  Lost Summoner Kitty   
4      761140   4.99   2018-01-04  Lost Summoner Kitty   
...       ...    ...          ...                  ...   
57577  610660   1.99   2018-01-04        Russian Roads   
57578  610660   1.99   2018-01-04        Russian Roads   
57579  610660   1.99   2018-01-04        Russian Roads   
57580  658870   4.99   2017-09-02  EXIT 2 - Directions   
57581  658870   4.99   2017-09-02  EXIT 2 - Directions   

                      developer  
0                     Kotoshiro  
1                     Kotoshiro  
2                     Kotoshiro  
3                     Kotoshiro  
4                     Kotoshiro  
...                         ...  
57577  Laush Dmitriy Sergeevich  
57578  Laush Dmitriy Sergeevich

ENDPOINT 1

In [11]:
data_games.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'item_id', 'developer'],
      dtype='object')

In [28]:
data_games = pd.read_csv('limpio_games_gituno.csv')
data_games 

id  price release_date             app_name  \
0      761140   4.99   2018-01-04  Lost Summoner Kitty   
1      761140   4.99   2018-01-04  Lost Summoner Kitty   
2      761140   4.99   2018-01-04  Lost Summoner Kitty   
3      761140   4.99   2018-01-04  Lost Summoner Kitty   
4      761140   4.99   2018-01-04  Lost Summoner Kitty   
...       ...    ...          ...                  ...   
57577  610660   1.99   2018-01-04        Russian Roads   
57578  610660   1.99   2018-01-04        Russian Roads   
57579  610660   1.99   2018-01-04        Russian Roads   
57580  658870   4.99   2017-09-02  EXIT 2 - Directions   
57581  658870   4.99   2017-09-02  EXIT 2 - Directions   

                      developer  
0                     Kotoshiro  
1                     Kotoshiro  
2                     Kotoshiro  
3                     Kotoshiro  
4                     Kotoshiro  
...                         ...  
57577  Laush Dmitriy Sergeevich  
57578  Laush Dmitriy Sergeevich  
57579  Laush Dmitriy Sergeevich  
57580             xropi,stev3ns  
57581             xropi,stev3ns  

[57582 rows x 5 columns]

In [29]:
import pandas as pd
import numpy as np
import json

data_games = pd.read_csv('limpio_games_gituno.csv')

def process_developer(desarrollador: str, data_games: pd.DataFrame):
    # Filtrar los elementos asociados a la empresa desarrolladora especificada
    elementos_desarrollador = data_games[data_games['developer'] == desarrollador]

    if elementos_desarrollador.empty:
        return "No se encontraron elementos para la empresa desarrolladora especificada."

    # Convert 'release_year' column to native Python integers
    elementos_desarrollador['release_year'] = elementos_desarrollador['release_date'].str.extract(r'(\d{4})')

    # Calcular la cantidad de elementos lanzados por año
    elementos_por_anio = elementos_desarrollador['release_year'].value_counts().reset_index()
    elementos_por_anio.columns = ['Anio', 'Cantidad de Items']

    # Calcular el porcentaje de elementos gratuitos por año
    elementos_gratuitos_por_anio = elementos_desarrollador[elementos_desarrollador['price'] == 0]
    elementos_gratuitos_por_anio = elementos_gratuitos_por_anio['release_year'].value_counts().reset_index()
    elementos_gratuitos_por_anio.columns = ['Anio', 'Contenido Free en %']

    # Fusionar ambos DataFrames para obtener el resultado final
    resultado = elementos_por_anio.merge(elementos_gratuitos_por_anio, on='Anio', how='left')
    resultado['Contenido Free en %'] = (resultado['Contenido Free en %'] / resultado['Cantidad de Items']) * 100

    # Reemplazar NaN, Infinity y -Infinity por None
    resultado = resultado.replace([np.inf, -np.inf, np.nan], None)

    # Convertir DataFrame a diccionario de Python
    resultado_dict = resultado.to_dict()

    # Convertir diccionario de Python a JSON
    resultado_json = json.dumps(resultado_dict, separators=(",", ":"))

    return resultado_json

# Ejemplo de uso:
desarrollador = 'Valve'
resultado_json = process_developer(desarrollador, data_games)
print(resultado_json)


{"Anio":{"0":"2014","1":"2004","2":"2007","3":"2012","4":"2017","5":"2010","6":"2011","7":"2006","8":"2000","9":"1998","10":"2009","11":"2008","12":"2005","13":"1999","14":"2001","15":"2003"},"Cantidad de Items":{"0":9,"1":5,"2":4,"3":3,"4":3,"5":2,"6":2,"7":2,"8":2,"9":1,"10":1,"11":1,"12":1,"13":1,"14":1,"15":1},"Contenido Free en %":{"0":100.0,"1":null,"2":50.0,"3":66.66666666666666,"4":null,"5":50.0,"6":null,"7":null,"8":null,"9":null,"10":null,"11":null,"12":100.0,"13":null,"14":null,"15":null}}


C:\Users\juana\AppData\Local\Temp\ipykernel_6252\3092447671.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  elementos_desarrollador['release_year'] = elementos_desarrollador['release_date'].str.extract(r'(\d{4})')


ACQUIRE Corp.

773

KintoGames

Noah System,Mgame

Fistful of Frags Team

S2 Games, LLC

Galactic Cafe

Odin Game Studio

MOONSTONE

Sluggerfly